File: integration_tests/dart_client_tests/test/mutations_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'package:graphlink_client_integration_tests/generated/enums/priority.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/create_user_input.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/update_user_input.dart';
import 'package:graphlink_client_integration_tests/generated/inputs/address_input.dart';
import 'fixtures.dart';
import 'mock_adapter.dart';

final _minimalInput = CreateUserInput(
  name: 'Alice Smith',
  email: 'alice@test.com',
  status: UserStatus.ACTIVE,
  address: AddressInput(street: '123 Main St', city: 'Springfield', country: 'US'),
);

void main() {
  late MockAdapter adapter;
  late GraphLinkClient client;

  setUp(() {
    adapter = MockAdapter();
    client = GraphLinkClient(adapter: adapter.call, wsAdapter: MockWebSocketAdapter());
  });

  // ── createUser — response deserialization ─────────────────────────────────

  group('createUser — scalar fields', () {
    setUp(() => adapter.registerData('createUser', {'createUser': kUserAliceJson}));

    test('id is deserialized correctly', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.id, equals('user-1'));
    });

    test('name is deserialized correctly', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.name, equals('Alice Smith'));
    });

    test('email is deserialized correctly', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.email, equals('alice@test.com'));
    });

    test('status is deserialized as enum', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.status, equals(UserStatus.ACTIVE));
    });

    test('priority is deserialized as enum when present', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.priority, equals(Priority.HIGH));
    });
  });

  group('createUser — nested address', () {
    setUp(() => adapter.registerData('createUser', {'createUser': kUserAliceJson}));

    test('address.street is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.street, equals('123 Main St'));
    });

    test('address.city is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.city, equals('Springfield'));
    });

    test('address.country is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.country, equals('US'));
    });

    test('address.zip is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.address.zip, equals('12345'));
    });

    test('nullable billingAddress is null', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.billingAddress, isNull);
    });
  });

  group('createUser — list fields', () {
    setUp(() => adapter.registerData('createUser', {'createUser': kUserAliceJson}));

    test('tags list is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.tags, equals(['admin', 'beta']));
    });

    test('scores list is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.scores, equals([10, 20, 30]));
    });
  });

  group('createUser — response with Bob (nullable fields populated)', () {
    setUp(() => adapter.registerData('createUser', {'createUser': kUserBobJson}));

    test('status INACTIVE is deserialized', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.status, equals(UserStatus.INACTIVE));
    });

    test('nullable priority is null', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.priority, isNull);
    });

    test('billingAddress is deserialized when present', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.billingAddress, isNotNull);
      expect(res.createUser.billingAddress!.city, equals('Capital City'));
    });

    test('empty tags list is deserialized as empty list', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.tags, isEmpty);
    });

    test('nullable scores is null', () async {
      final res = await client.mutations.createUser(input: _minimalInput);
      expect(res.createUser.scores, isNull);
    });
  });

  group('createUser — outgoing request shape', () {
    setUp(() => adapter.registerData('createUser', {'createUser': kUserAliceJson}));

    test('operation name is createUser', () async {
      await client.mutations.createUser(input: _minimalInput);
      expect(adapter.lastCall!.operationName, equals('createUser'));
    });

    test('input variable is present', () async {
      await client.mutations.createUser(input: _minimalInput);
      expect(adapter.lastCall!.variables['input'], isNotNull);
    });
  });

  // ── deleteUser — scalar return + variable shape ───────────────────────────

  group('deleteUser — returns bool', () {
    test('returns true when server responds true', () async {
      adapter.registerData('deleteUser', {'deleteUser': true});
      final res = await client.mutations.deleteUser(id: 'user-1');
      expect(res.deleteUser, isTrue);
    });

    test('returns false when server responds false', () async {
      adapter.registerData('deleteUser', {'deleteUser': false});
      final res = await client.mutations.deleteUser(id: 'user-1');
      expect(res.deleteUser, isFalse);
    });
  });

  group('deleteUser — outgoing request shape', () {
    setUp(() => adapter.registerData('deleteUser', {'deleteUser': true}));

    test('id variable is sent correctly', () async {
      await client.mutations.deleteUser(id: 'user-99');
      expect(adapter.lastCall!.variables['id'], equals('user-99'));
    });

    test('operation name is deleteUser', () async {
      await client.mutations.deleteUser(id: 'user-1');
      expect(adapter.lastCall!.operationName, equals('deleteUser'));
    });
  });

  // ── updateUser — response deserialization + variable shape ────────────────

  group('updateUser — response deserialization', () {
    setUp(() => adapter.registerData('updateUser', {'updateUser': kUserAliceJson}));

    test('returns a User with correct id', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Alice Smith'),
      );
      expect(res.updateUser.id, equals('user-1'));
    });

    test('returns a User with correct name', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Alice Smith'),
      );
      expect(res.updateUser.name, equals('Alice Smith'));
    });

    test('status is deserialized as enum', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(status: UserStatus.ACTIVE),
      );
      expect(res.updateUser.status, equals(UserStatus.ACTIVE));
    });

    test('nested address is deserialized', () async {
      final res = await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Alice Smith'),
      );
      expect(res.updateUser.address.street, equals('123 Main St'));
    });
  });

  group('updateUser — outgoing request shape', () {
    setUp(() => adapter.registerData('updateUser', {'updateUser': kUserAliceJson}));

    test('id is passed as top-level variable', () async {
      await client.mutations.updateUser(
        id: 'user-42',
        input: UpdateUserInput(name: 'Updated'),
      );
      expect(adapter.lastCall!.variables['id'], equals('user-42'));
    });

    test('input variable is present alongside id', () async {
      await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Updated'),
      );
      expect(adapter.lastCall!.variables['input'], isNotNull);
    });

    test('operation name is updateUser', () async {
      await client.mutations.updateUser(
        id: 'user-1',
        input: UpdateUserInput(name: 'Updated'),
      );
      expect(adapter.lastCall!.operationName, equals('updateUser'));
    });
  });
}

File: integration_tests/typescript_client_tests/test/mutations.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import { Priority } from '../lib/generated/enums/priority.js';
import type { CreateUserInput } from '../lib/generated/inputs/create-user-input.js';
import type { UpdateUserInput } from '../lib/generated/inputs/update-user-input.js';
import { MockAdapter, MockWsAdapter } from './mock-adapter.ts';
import { kUserAliceJson, kUserBobJson } from './fixtures.ts';

const minimalInput: CreateUserInput = {
  name: 'Alice Smith',
  email: 'alice@test.com',
  status: UserStatus.Active,
  address: { street: '123 Main St', city: 'Springfield', country: 'US' },
};

let adapter: MockAdapter;
let client: GraphLinkClient;

beforeEach(() => {
  adapter = new MockAdapter();
  client = new GraphLinkClient(adapter.call, new MockWsAdapter());
});

// ── createUser — response deserialization ─────────────────────────────────────

describe('createUser — scalar fields', () => {
  beforeEach(() => { adapter.registerData('createUser', { createUser: kUserAliceJson }); });

  it('id is deserialized correctly', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.id).toBe('user-1');
  });

  it('name is deserialized correctly', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.name).toBe('Alice Smith');
  });

  it('email is deserialized correctly', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.email).toBe('alice@test.com');
  });

  it('status is deserialized as enum', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.status).toBe(UserStatus.Active);
  });

  it('priority is deserialized as enum when present', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.priority).toBe(Priority.High);
  });
});

describe('createUser — nested address', () => {
  beforeEach(() => { adapter.registerData('createUser', { createUser: kUserAliceJson }); });

  it('address.street is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.address.street).toBe('123 Main St');
  });

  it('address.city is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.address.city).toBe('Springfield');
  });

  it('address.country is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.address.country).toBe('US');
  });

  it('address.zip is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.address.zip).toBe('12345');
  });

  it('nullable billingAddress is null', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.billingAddress).toBeNull();
  });
});

describe('createUser — list fields', () => {
  beforeEach(() => { adapter.registerData('createUser', { createUser: kUserAliceJson }); });

  it('tags list is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.tags).toEqual(['admin', 'beta']);
  });

  it('scores list is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.scores).toEqual([10, 20, 30]);
  });
});

describe('createUser — response with Bob (nullable fields populated)', () => {
  beforeEach(() => { adapter.registerData('createUser', { createUser: kUserBobJson }); });

  it('status INACTIVE is deserialized', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.status).toBe(UserStatus.Inactive);
  });

  it('nullable priority is null', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.priority).toBeNull();
  });

  it('billingAddress is deserialized when present', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.billingAddress).not.toBeNull();
    expect(res.createUser.billingAddress!.city).toBe('Capital City');
  });

  it('empty tags list is deserialized as empty list', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.tags).toHaveLength(0);
  });

  it('nullable scores is null', async () => {
    const res = await client.mutations.createUser({ input: minimalInput });
    expect(res.createUser.scores).toBeNull();
  });
});

describe('createUser — outgoing request shape', () => {
  beforeEach(() => { adapter.registerData('createUser', { createUser: kUserAliceJson }); });

  it('operation name is createUser', async () => {
    await client.mutations.createUser({ input: minimalInput });
    expect(adapter.lastCall!.operationName).toBe('createUser');
  });

  it('input variable is present', async () => {
    await client.mutations.createUser({ input: minimalInput });
    expect(adapter.lastCall!.variables['input']).not.toBeNull();
  });
});

// ── deleteUser ────────────────────────────────────────────────────────────────

describe('deleteUser — returns bool', () => {
  it('returns true when server responds true', async () => {
    adapter.registerData('deleteUser', { deleteUser: true });
    const res = await client.mutations.deleteUser({ id: 'user-1' });
    expect(res.deleteUser).toBe(true);
  });

  it('returns false when server responds false', async () => {
    adapter.registerData('deleteUser', { deleteUser: false });
    const res = await client.mutations.deleteUser({ id: 'user-1' });
    expect(res.deleteUser).toBe(false);
  });
});

describe('deleteUser — outgoing request shape', () => {
  beforeEach(() => { adapter.registerData('deleteUser', { deleteUser: true }); });

  it('id variable is sent correctly', async () => {
    await client.mutations.deleteUser({ id: 'user-99' });
    expect(adapter.lastCall!.variables['id']).toBe('user-99');
  });

  it('operation name is deleteUser', async () => {
    await client.mutations.deleteUser({ id: 'user-1' });
    expect(adapter.lastCall!.operationName).toBe('deleteUser');
  });
});

// ── updateUser ────────────────────────────────────────────────────────────────

describe('updateUser — response deserialization', () => {
  beforeEach(() => { adapter.registerData('updateUser', { updateUser: kUserAliceJson }); });

  it('returns a User with correct id', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Alice Smith' } as UpdateUserInput });
    expect(res.updateUser.id).toBe('user-1');
  });

  it('returns a User with correct name', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Alice Smith' } as UpdateUserInput });
    expect(res.updateUser.name).toBe('Alice Smith');
  });

  it('status is deserialized as enum', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { status: UserStatus.Active } });
    expect(res.updateUser.status).toBe(UserStatus.Active);
  });

  it('nested address is deserialized', async () => {
    const res = await client.mutations.updateUser({ id: 'user-1', input: { name: 'Alice Smith' } as UpdateUserInput });
    expect(res.updateUser.address.street).toBe('123 Main St');
  });
});

describe('updateUser — outgoing request shape', () => {
  beforeEach(() => { adapter.registerData('updateUser', { updateUser: kUserAliceJson }); });

  it('id is passed as top-level variable', async () => {
    await client.mutations.updateUser({ id: 'user-42', input: { name: 'Updated' } as UpdateUserInput });
    expect(adapter.lastCall!.variables['id']).toBe('user-42');
  });

  it('input variable is present alongside id', async () => {
    await client.mutations.updateUser({ id: 'user-1', input: { name: 'Updated' } as UpdateUserInput });
    expect(adapter.lastCall!.variables['input']).not.toBeNull();
  });

  it('operation name is updateUser', async () => {
    await client.mutations.updateUser({ id: 'user-1', input: { name: 'Updated' } as UpdateUserInput });
    expect(adapter.lastCall!.operationName).toBe('updateUser');
  });
});
